### PR TITLE
Feature/auth

### PR DESCRIPTION
**Pull request by RPGPH**
Draft for authentication and test discussion

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -43,39 +32,93 @@ export type AcmeGroup = Opaque<any, 'AcmeGroup'>;
...
+        .filter((item) => item.search('JSESSIONID') >= 0)[0] || ''
+    );
+  }
+
+  public async logout(cookie: string): Promise<void> {
+    const logoutRoute = this.withBaseUri('logout');
+    await this.getRequest(logoutRoute, cookie);
+  }
+
+  public async verifyAuthentication(): Promise<void> {
```
Comment (RPGPH) [inline src/client.ts:102]: Auth verification flow:

login() -> test cookie against an api (proceed if successful, else throw Auth error) -> logout()

File: src/client.ts [hunk 2]
```diff
@@ -43,39 +32,93 @@ export type AcmeGroup = Opaque<any, 'AcmeGroup'>;
...
+    const cookie = await this.login();
+
+    // NOTE: This only works with 'System admin' role
+    // no API endpoint is available for all roles, need to find
+    // a better endpoint to test the cookie that is not hindered
+    // by role permissions.
```
Comment (RPGPH) [inline src/client.ts:106]: Permissions table: https://nest.cybereason.com/documentation/api-documentation/all-versions/required-permissions

A false authentication error would occur when the user has no permission to the testing endpoint if we use a try-catch statement (moving away from it for now). We need an endpoint that is available for all roles but no such thing exists.

File: src/client.ts [hunk 3]
```diff
@@ -43,39 +32,93 @@ export type AcmeGroup = Opaque<any, 'AcmeGroup'>;
...
+    const groupsUri = this.withBaseUri('rest/groups');
+    const res = await this.getRequest(groupsUri, cookie);
+
+    // if cookie is invalid, API responds with login.html
+    if (res.headers.get('content-type') === 'text/html')
+      throw new IntegrationProviderAuthenticationError({
+        endpoint: groupsUri,
+        status: res.status,
+        statusText: res.statusText,
+      });
```
Comment (RPGPH) [inline src/client.ts]: Problem: API doesn't respond with a 401 error when the cookie is invalid. Hacked a way to check if the response content is HTML instead which shouldn't happen when the cookie is valid. Feels like a bad solution. Is this ok?

See supported error codes here:
https://nest.cybereason.com/documentation/api-documentation/all-versions/responses#error-codes

File: src/validateInvocation.test.ts [hunk 4]
```diff
@@ -30,12 +30,16 @@ it('auth error', async () => {
...
+      cybereasonHost: 'INVALID',
+      cybereasonPort: 'INVALID',
     },
   });
 
+  // TEMP: Change assertion pass test and push for draft PR
+
   await expect(validateInvocation(executionContext)).rejects.toThrow(
-    IntegrationProviderAuthenticationError,
+    IntegrationProviderAPIError,
```
Comment (RPGPH) [inline src/validateInvocation.test.ts:42]: If the host and port is invalid, it will throw an API error instead because you can't verify authentication without a valid host and port. Is this acceptable or is it required to assert `IntegrationProviderAuthenticationError` here, and if so, how do we approach it without pushing the server's credentials?

File: test/config.ts [hunk 5]
```diff
@@ -7,10 +7,17 @@ if (process.env.LOAD_ENV) {
...
 export const integrationConfig: IntegrationConfig = {
-  clientId: process.env.CLIENT_ID || DEFAULT_CLIENT_ID,
-  clientSecret: process.env.CLIENT_SECRET || DEFAULT_CLIENT_SECRET,
+  cybereasonId: process.env.CYBEREASON_ID || DEFAULT_CYBEREASON_ID,
+  cybereasonPassword:
+    process.env.CYBEREASON_SECRET || DEFAULT_CYBEREASON_SECRET,
+  cybereasonHost:
+    process.env.DEFAULT_CYBEREASON_HOST || DEFAULT_CYBEREASON_HOST,
+  cybereasonPort:
+    process.env.DEFAULT_CYBEREASON_PORT || DEFAULT_CYBEREASON_PORT,
```
Comment (RPGPH) [inline test/config.ts:22]: Host and port is user dependent. Is this acceptable?

**Comment by RPGPH** [comment]
Hello Stefan @eXtremeX,

I'm running into some problems with authentication verification and I figured I should ask for your opinion about some of them. Feel free to look at the comments. :)

File: src/client.ts (hunk 1 above)
Comment (RPGPH) [reply src/client.ts:102]: Note: This would eventually be done on every API request since as discussed, we can't use setData. Will probably move this to a reusable method down the line.

Comment (eXtremeX) [reply src/client.ts:102]: This is a good direction! Here's a suggestion:

1. Create something like `private static authCookie: string` as a Class' member.
2. Then your `login()` method will have a function of getting the auth cookie, a better name might be `getAuthCookie()`
3. In your request methods (`getRequest()`, `postRequest()` *), you can check if the cookie is already fetched before your `fetch()` call:
```
if (!APIClient.authCookie) {
  await this.getAuthCookie();
}
```

Since every call will go through the request method, its responsibility will be to check if the cookie exists and if not to fetch it. What do you think about this? @RPGPH 

Here's a good example of this in action:
1) https://github.com/JupiterOne/graph-checkmarx/blob/master/src/client.ts#L33 (Also please have a similar comment)
2) https://github.com/JupiterOne/graph-checkmarx/blob/master/src/client.ts#L65
3) https://github.com/JupiterOne/graph-checkmarx/blob/master/src/client.ts#L50

*) Also, can we have a single `request` method that can do both POST and GET requests instead of two, as seen here: https://github.com/JupiterOne/graph-checkmarx/blob/master/src/client.ts#L47?

File: src/client.ts (hunk 2 above)
Comment (eXtremeX) [reply src/client.ts:106]: Yep, I see, thanks for pointing this out. Let's revisit this after you've had the chance to interact with more endpoints.

File: src/client.ts (hunk 3 above)
Comment (eXtremeX) [reply src/client.ts]: Noted! It does feel hacky but sometimes there are edge cases such as this one. Let's keep this one for now and later if we can think of some neater way of handling this, we'll change it :+1:

File: test/config.ts (hunk 5 above)
Comment (eXtremeX) [reply test/config.ts:22]: Yes, the end-user/customer will need to provide to the integration their host and port :100:.

File: src/validateInvocation.test.ts (hunk 4 above)
Comment (eXtremeX) [reply src/validateInvocation.test.ts:42]: Makes sense! Just add that to the comment somewhere so that we note down the reason behind this change.

**Review by eXtremeX — CHANGES_REQUESTED** [review]

File: src/client.ts [hunk 6]
```diff
@@ -43,39 +32,88 @@ export type AcmeGroup = Opaque<any, 'AcmeGroup'>;
...
-        },
-        (res) => {
-          if (res.statusCode !== 200) {
-            reject(new Error('Provider authentication failed'));
-          } else {
-            resolve();
-          }
-        },
-      );
+  private accessToken: string;
```
Comment (RPGPH) [inline src/client.ts]: TEMP: Leaving it as a non static variable for the meantime because we are getting a new `accessToken` per request.

https://github.com/JupiterOne/graph-checkmarx/blob/5301833e04eca44cf0df303d68a79525bd4e549a/src/client.ts#L33

The problem I'm having with this approach is after the integration steps have been done, the generated `accessToken` remains valid. It feels like it might be a security concern. I will make it a `static` class if you guys are okay with leaving it as is or if we find a way to run a `logout()` call after integration.

Comment (eXtremeX) [reply src/client.ts]: Non-static is fine :+1: , it'd just mean that each step will get a new cookie (e.g. each step instantiates a new client and it gets a new cookie), so long the Cybereason doesn't mind a single account requesting many cookies at "once", it should work. Let's try it and see.

P.S I'd still name with some name that has the word 'cookie' in it, the value in the example you were referencing was indeed the access_token, hence that name :slightly_smiling_face:
